kindelia_core/src/builder.rs

```diff
@@ -0,0 +1,227 @@
+  #[error("Missing required field {field}")]
+  MissingRequired { field: String },
+}
+
+/// A Builder to simplify creation of a Node
+///
+/// Example:
+///
+///   let (node_query_sender, node) = NodeBuilder::default()
+///    .network_id(network_id)
+///    .comm(comm)
+///    .miner_comm(miner_comm)
+///    .addr(addr)
+///    .storage(file_writer)
+///    .genesis_code(genesis_code)
+///    .data_path(data_path)
+///    .build()?;
+pub struct NodeBuilder<C: ProtoComm, S: BlockStorage> {
+  network_id: u32,          // Network ID / magic number
+  comm: Option<C>,          // UDP socket
+  addr: Option<C::Address>, // UDP port
+  storage: Option<S>,       // A `BlockStorage` implementation
+  peers: Vec<C::Address>,   // peers store and state control
+
+  #[cfg(feature = "events")]
+  event_emitter: Option<mpsc::Sender<NodeEventEmittedInfo>>,
+  miner_comm: Option<MinerCommunication>,
+
+  // these do not exist in Node. they are "transient"
+  data_path: PathBuf, // path to ~/.kindelia/state/<network_id>
+  genesis_code: String, // statements in genesis block body
+}
+
+impl<C: ProtoComm, S: BlockStorage> NodeBuilder<C, S> {
+  /// set network identifier
+  pub fn network_id(mut self, network_id: u32) -> Self {
+    self.network_id = network_id;
+    self
+  }
+
+  /// set comm channel, eg std::net::UdpSocket + ProtoComm
+  pub fn comm(mut self, comm: C) -> Self {
+    self.comm = Some(comm);
+    self
+  }
+
+  /// set comm addr, eg net::Address
+  pub fn addr(mut self, addr: C::Address) -> Self {
+    self.addr = Some(addr);
+    self
+  }
+
+  /// set storage, eg persistence::SimpleFileStorage
+  pub fn storage(mut self, storage: S) -> Self {
+    self.storage = Some(storage);
+    self
+  }
+
+  /// set source code statements for the genesis block
+  pub fn genesis_code(mut self, genesis_code: String) -> Self {
+    self.genesis_code = genesis_code;
+    self
+  }
+
+  /// set data path, eg ~/.kindelia/state/<network_id>
+  pub fn data_path(mut self, data_path: PathBuf) -> Self {
+    self.data_path = data_path;
+    self
+  }
+
+  /// set initial peers for this node to contact
+  pub fn peers(mut self, peers: Vec<C::Address>) -> Self {
+    self.peers = peers;
+    self
+  }
+
+  /// set event emitter
+  pub fn event_emitter(
+    mut self,
+    emitter: mpsc::Sender<NodeEventEmittedInfo>,
+  ) -> Self {
+    self.event_emitter = Some(emitter);
+    self
+  }
+
+  /// set miner communication channel, or None if not mining.
+  pub fn miner_comm(mut self, miner_comm: Option<MinerCommunication>) -> Self {
+    self.miner_comm = miner_comm;
+    self
+  }
+
+  /// validates builder inputs and builds a Node instance.
+  //
+  // note: I would like to define a type alias to get rid of this clippy
+  //       warning however it appears to require "inherent associated types"
+  //       due to the use of generics, which are unstable.
+  //       see: https://github.com/rust-lang/rust/issues/8995
+  #[allow(clippy::type_complexity)]
```

Comment:
I think it would be possible to make a type out of `impl NodeBuilder` like this:

```
pub type BuilderReturn<C, S> = (mpsc::SyncSender<NodeRequest<C>>, Node<C, S>);
```

and use it as build function return like: `Result<BuilderReturn<C, S>, NodeBuilderError>`

(The `BuilderReturn` name is just an example)